### ISSUE TITLE
Implement health check for the tracker application

### ISSUE THREAD

**Issue by josecelano** [torrust/torrust-tracker#508]
Implement a HEALTHCHECK for the whole application. [We are doing the same for the index](https://github.com/torrust/torrust-index/issues/254). 

The goal is to add a line similar to this:

```dockerfile
HEALTHCHECK --interval=5s --timeout=5s --start-period=3s --retries=3 \  
  CMD    /usr/bin/health_check --http http://localhost:${API_PORT}/health_check \
      && /usr/bin/health_check --http http://localhost:${HTTP_PORT}/health_check \
      && /usr/bin/health_check --udp udp://localhost:${UDP_PORT} \
      || exit 1
````

So when you run the tracker in a container, it can show its status (healthy or not).

For the HTTP tracker and Tracker API, we can add `health_check` endpoints like in the [Index](https://github.com/torrust/torrust-index/issues/254).

For the UDP Tracker, maybe we can try to connect.

**Comment by josecelano** [comment]
Today @da2ce7 and I agree on try this approach:

- We add a high-level Management API.
- That API uses a new socket.
- For the time being, it only has one endpoint: `GET HTTP://localhost:1313/health_check`.
- The health check handler has to check if the other services are healthy.

### Proposed solution 1: one-shot channels

The proposed implementation is to use channels between the main Management API and the other three services: API, HTTP Tracker and UDP Tracker. We can modify the `start_job` functions to accept a new parameter which is the channel sender to send a message:

```rust
use std::sync::mpsc;

let (tx_tracker_api, rx_tracker_api) = mpsc::channel();
let (tx_http_tracker, rx_http_tracker) = mpsc::channel();
let (tx_udp_tracker, rx_udp_tracker) = mpsc::channel();
```

When we start the service we send back to the main app a message and we change the state so that the Management API can use that state. With this solution, this would be checked only once. If the service crashes the health checkpoint is going to return always a healthy state.

We can send the message from the service, not from the launcher, but in that case, the service must know how to communicate with the high-level service. And that is something I do not like. We should be able to move the UDP to its own package in the future.

### Proposed solution 2: Rest API (polling)

What I did for the Index was [this](https://github.com/torrust/torrust-index/blob/develop/src/console/tracker_statistics_importer.rs#L85-L87):

There are 2 services:

- The API
- A runner (importer executed every 2 minutes)

The importer process sends a heartbeat periodically to the importer API to inform it's still alive. Both components are inside the same package/module. The high-level service polls both low-level services, so neither the API nor the importer know the heath checker.

We can do the same here. Adding an endpoint for health checks to the three services would allow us to poll instead of pushing from those services. I think Rest API are a less couple mechanism in comparison with channels if we want to create separate packages for the UDP Tracker, HTTP tracker and API.

In summary, service should provide a way to monitor them without coupling them to the processes that are monitoring. We could implement a "push" approach, where the services send a message periodically to the monitor but I would use anyway an API instead of shared memory between processes. With shared memory, we will never be able to deploy those services independently. And there could be some scenarios where we want to do that. Probably API requests are less than UDP announce requests.

@da2ce7 what do you think?

**Comment by josecelano** [comment]
Hi @da2ce7, I've just realized that:

- You can use port 0 for all services' sockets (API, HTTP Trcker and UDP Tracker).
- You can have more than one UDP tracker and HTTP tracker.

So the health check endpoint should check that all services (1 API, N UDP trackers and N HTTP trackers) are running and listening on their ports. We need to pass the socket port to this service and that means we have to make a lot of changes. Some needed changes were already defined in these issues:

- https://github.com/torrust/torrust-tracker/issues/274
- https://github.com/torrust/torrust-tracker/issues/186

We have to implement those issues and also pass the port numbers not only to the launcher (parent thread when you launch the service) but also the main app so the main app can pass the ports to the Health Check Endpoint.

What I'm going to do is just check services when the port number is not 0 because I only need to inject the configuration. In the future, when we implement the other issues (and we have the port numbers) we can check all of them.

I think the complete version (including ports 0) is out of scope for the time being. We wanted to add the `HEALTHCHECK` instruction to the `Containerfile`. If you define the ports the health check is going to check them all. If you use port 0 for all services, the health check is not going to check any service but at least is checking that the main app did not crashed, and that would be also OK for our purposes. In fact, I'm starting to change my mind about this approach. Maybe we should only check that the main container process is OK. But I think it's an good feature to have becuase when you run the app in containerized env containers should not receive request is they are not ready. I mean, I not only want to check that is running but that is ready to accept requests.
